tensorflow_probability/python/distributions/johnson_su.py

```diff
@@ -0,0 +1,335 @@
+
+  The probability density function (pdf) is,
+
+  ```none
+  pdf(x; gamma, delta, xi, sigma) = exp(-0.5 (gamma + delta arcsinh(y))**2) / Z
+  where,
+  y = (x - xi) / sigma
+  Z = sigma sqrt(2 pi) sqrt(1 + y**2) / delta
+  ```
+
+  where:
+  * `loc = xi`,
+  * `scale = sigma`, and,
+  * `Z` is the normalization constant.
+
+  The JohnsonSU distribution is a member of the [location-scale family](
+  https://en.wikipedia.org/wiki/Location-scale_family), i.e., it can be
+  constructed as,
+
+  ```none
+  X ~ JohnsonSU(gamma, delta, loc=0, scale=1)
+  Y = loc + scale * X
+  ```
+
+  #### Examples
+
+  Examples of initialization of one or a batch of distributions.
+
+  ```python
+  import tensorflow_probability as tfp
+  tfd = tfp.distributions
+
+  # Define a single scalar Johnson's SU-distribution.
+  single_dist = tfd.JohnsonSU(gamma=-2., delta=2., loc=1.1, scale=1.5)
+
+  # Evaluate the pdf at 1, returning a scalar Tensor.
+  single_dist.prob(1.)
+
+  # Define a batch of two scalar valued Johnson SU's.
+  # The first has shape parameters 1 and 2, mean 3, and scale 11.
+  # The second 4, 5, 6 and 22.
+  multi_dist = tfd.JohnsonSU(gamma=[1, 4], delta=[2, 5],
+                             loc=[3, 6], scale=[11, 22.])
+
+  # Evaluate the pdf of the first distribution on 0, and the second on 1.5,
+  # returning a length two tensor.
+  multi_dist.prob([0, 1.5])
+
+  # Get 3 samples, returning a 3 x 2 tensor.
+  multi_dist.sample(3)
+  ```
+
+  Arguments are broadcast when possible.
+
+  ```python
+  # Define a batch of two Johnson's SU distributions.
+  # Both have gamma 2, delta 3 and mean 1, but different scales.
+  dist = tfd.JohnsonSU(gamma=2, delta=3, loc=1, scale=[11, 22.])
+
+  # Evaluate the pdf of both distributions on the same point, 3.0,
+  # returning a length 2 tensor.
+  dist.prob(3.0)
+  ```
+
+  Compute the gradients of samples w.r.t. the parameters:
+
+  ```python
+  gamma = tf.Variable(2.0)
+  delta = tf.Variable(3.0)
+  loc = tf.Variable(2.0)
+  scale = tf.Variable(11.0)
+  dist = tfd.JohnsonSU(gamma=gamma, delta=delta, loc=loc, scale=scale)
+  with tf.GradientTape() as tape:
+    samples = dist.sample(5)  # Shape [5]
+    loss = tf.reduce_mean(tf.square(samples))  # Arbitrary loss function
+  # Unbiased stochastic gradients of the loss function
+  grads = tape.gradient(loss, dist.variables)
+  ```
+
+  """
+
+  def __init__(self,
+               gamma,
+               delta,
+               loc,
+               scale,
+               validate_args=False,
+               allow_nan_stats=True,
+               name=None):
+    """Construct Johnson's SU distributions.
+
+    The distributions have shape parameteres `delta` and `gamma`, mean `loc`,
+    and scale `scale`.
+
+    The parameters `delta`, `gamma`, `loc`, and `scale` must be shaped in a way
+    that supports broadcasting (e.g. `gamma + delta + loc + scale` is a valid
+    operation).
+
+    Args:
+      gamma: Floating-point `Tensor`. The shape parameter(s) of the
```

Comment:
Are there more informative names for `gamma` and `delta` that we could add? Generally for TFP, we try to avoid greek names, and try to give names that correspond to the properties (e.g. for `Beta` we use `concentration0` and `concentration1` to describe the concentration parameters which specify how the mass is weighted near `0` and `1`.

Reply:
I saw `gamma`/`delta` and `a`/`b` names in other places. Do you have suggestions on what can be used here?

Reply:
I think we can reuse `tailweight` for `delta` and `skewness` for `gamma` (copied from SinhArcsinh bijector). `delta` controls how fat/thin the tails are, while `gamma` shifts the mode/center of the distribution, which corresponds to skewness (which can be seen from plots).

The only weird thing is that the actual skewness of the distribution is going to be a function of `gamma / delta` most likley, but I am fine naming this `skewness` since it matches the bijector.

Reply:
Renamed